StartTLS implementation

kozzztik opened the pull request:
Implementation of STARTLS extention. Well tested in production and real usage. Works only on python 3.5 (tested on 3.5.2). On python < 3.4 and without SSL support enabled will simply refuse STARTTLS command with right message.

warsaw commented on aiosmtpd/smtp.py:
```diff
@@ -35,6 +42,8 @@ def __init__(self, handler,
                  hostname=None,
+                 tls_context=None,
+                 require_starttls=False,
                  loop=None):
```
I wonder if we should make `loop` a required keyword argument?  We keep inserting new parameters just before it, and that technically breaks the API.  Of course, we're still alpha so it's not that big of a deal, but what do you think of doing something like `require_starttls=False, *, loop=None` (suitably formatted)?

Then as part of the permanent API, if you specify `loop` you must do so as a keyword argument, and adding more positional arguments before it won't break the API.

warsaw commented on aiosmtpd/smtp.py:
```diff
@@ -69,6 +78,16 @@ def __init__(self, handler,
             self.hostname = socket.getfqdn()
+        self.tls_context = tls_context
+        if tls_context:
+            # through rfc3207 part 4.1 certificate checking is part of SMTP
```
Please capitalize "through".

warsaw commented on aiosmtpd/smtp.py:
```diff
@@ -78,13 +97,38 @@ def max_command_size_limit(self):
-        self._handler_coroutine = self.loop.create_task(self._handle_client())
+        if(self.transport is not None
+                and isinstance(transport, sslproto._SSLProtocolTransport)):
+            # It is STARTTLS connection over normal connection
```
Please end the sentence in a period.

warsaw commented on aiosmtpd/smtp.py:
```diff
@@ -78,13 +97,38 @@ def max_command_size_limit(self):
+            self.transport = transport
+            # Reset state due to rfc3207 part 4.2
+            self._set_rset_state()
+            self.seen_greeting = False
```
Why set `seen_greeting` here to False when the constructor already sets it to the empty string?

warsaw commented on aiosmtpd/smtp.py:
```diff
@@ -78,13 +97,38 @@ def max_command_size_limit(self):
+            # Why _extra is protected attribute?
+            extra = self._tls_protocol._extra
+            if hasattr(self.event_handler, 'handle_tls_handshake'):
+                self._tls_handshake_failed = \
```
Is there a way to refactor this (possibly using a local variable) to avoid the backslash?

Also, I often find double negatives to be difficult to reason about.  For example `self._tls_handshake_failed=False` means the the handshake succeeded right?  So why not name it `_tls_handshake_succeeded` and a True value is easier to parse mentally?

kozzztik replied:
Thats backslash is becouse of meaning ov variable. This is a flag to disable any server answer on SMTP commands after failed handshake. When it is false it means that there was no failed handshake. Also that mean that may be there was no handshake. I can add additional field to detect successful handshake if you want it.
I`ll fix other comments at Monday. Thanks for review! I temporary work without review of my code and that makes me too relaxed. Merging makes it worse.

warsaw replied:
About the backslash; I just meant that you might be able to use an extra local variable with a shorter name to avoid the backslash.  With the comment about changing the sense of the variable (to avoid the double negative), you might end up with something like this:

```
            if hasattr(self.event_handler, 'handle_tls_handshake'):
                handshake = self.event_handler.handle_tls_handshake(
                    extra['ssl_object'],
                    extra['peercert'],
                    extra['cipher'],
                    )
                self._tls_handshake_succeeded = handshake
````

warsaw commented on aiosmtpd/smtp.py:
```diff
@@ -78,13 +97,38 @@ def max_command_size_limit(self):
+                        extra['ssl_object'],
+                        extra['peercert'],
+                        extra['cipher'],
+                    )
```
Project style is to indent the closing parenthesis to under the previous line's first non-whitespace character.  Can you indent this (and following closing parens) 4 more spaces?

warsaw commented on aiosmtpd/smtp.py:
```diff
@@ -146,6 +190,16 @@ def _handle_client(self):
             if len(line) > max_sz:
                 yield from self.push('500 Error: line too long')
                 continue
+            if self._tls_handshake_failed and (command != 'QUIT'):
```
I don't think you need the parentheses here.

warsaw commented on aiosmtpd/smtp.py:
```diff
@@ -207,6 +263,35 @@ def smtp_QUIT(self, arg):
+        if arg:
+            yield from self.push('501 Syntax: STARTTLS')
+            return
+        if (not self.tls_context) or (not _has_ssl):
```
You don't need the parentheses here either.  Do you think it reads better though to write it

`if not (self.tls_context and _has_tls):`

?

warsaw commented on aiosmtpd/smtp.py:
```diff
@@ -207,6 +263,35 @@ def smtp_QUIT(self, arg):
+            return
+        if (not self.tls_context) or (not _has_ssl):
+            yield from self.push(
+                '454 TLS not available due to temporary reason'
```
Why the temporary failure here and the message?  Is it expected that if the client tries again in the future, `STARTTLS` will succeed?  I think it would be good to document the reasoning here in a comment.

warsaw commented on aiosmtpd/smtp.py:
```diff
@@ -207,6 +263,35 @@ def smtp_QUIT(self, arg):
+            server_side=True
+        )
+        # reconfigure transport layer
+        socket_transport = self.transport
```
Indent paren 4 spaces.

Please use complete sentences in the comments (i.e. start with a capital and end in a period).  Also for subsequent comments in this chunk.

warsaw commented on aiosmtpd/tests/test_starttls.py:
```diff
@@ -0,0 +1,219 @@
+    tls_context = ssl.create_default_context(ssl.Purpose.CLIENT_AUTH)
+    path = os.path.dirname(__file__)
+    path = os.path.join(path, 'certs')
+    tls_context.load_cert_chain(
```
Please use `pkg_resources` to do this instead of `__file__`.  The former module makes accessing files relative to the import path immune to packaging problems.  E.g.

```
path = pkg_resources.resource_filename('aiosmtpd.tests.certs', 'server.crt')
```

It means you'll have to put an empty `__init__.py` in the `certs` directory to make it a subpackage, but that's fine.

warsaw commented on aiosmtpd/tests/test_starttls.py:
```diff
@@ -0,0 +1,219 @@
+
+class TestStartTLS(unittest.TestCase):
+    if _has_ssl:
+        def test_starttls(self):
```
I think it would be better to just skip these tests if `_has_ssl` is false.  Better to report them as skipped than to just not execute them at all.  E.g. use `unittest.skipIf(not _has_ssl, 'SSL is not available')`

warsaw commented on aiosmtpd/tests/test_starttls.py:
```diff
@@ -0,0 +1,219 @@
+                self.assertEqual(code, 554)
+
+    else:
+        def test_starttls_fails_with_no_ssl(self):
```
Can this test be executed even when SSL is available?  Perhaps with a mock?

warsaw commented on aiosmtpd/tests/test_starttls.py:
```diff
@@ -0,0 +1,219 @@
+        self.addCleanup(controller.stop)
+        with SMTP(controller.hostname, controller.port) as client:
+            client.ehlo('example.com')
+            code, response = client.docmd("STARTTLS")
```
Can you use single quotes by default wherever possible?

warsaw commented on aiosmtpd/tests/test_starttls.py:
```diff
@@ -0,0 +1,219 @@
+            self.assertEqual(code, 501)
+
+if _has_ssl:
+    class TestTLSForgetsSessionData(unittest.TestCase):
```
Similarly you can `@unittest.skipIf()` for classes too, but do consider whether any of these tests are possible or useful when SSL is not available.

kozzztik commented:
Updated. Seems its needed to decrease py34 coverage requirements as it STARTLS is unavaliable and can`t be covered on 3.4.

warsaw commented:
Thanks for making the changes.  I'll review those again asap, but just a heads up that once #26 lands, it should at least fix the qa tests, so a rebase will be necessary.

warsaw commented:
On Nov 29, 2016, at 05:52 AM, Konstantin Volkov wrote:

>Updated. Seems its needed to decrease py34 coverage requirements as it
>STARTLS is unavaliable and can`t be covered on 3.4.

If needed, we can do that through some crafty tox/coverage.ini tricks without
decreasing coverage for 3.5 and 3.6.  I've done this in lots of projects.  I'm
happy to take that on, perhaps as a branch to be merged into your branch.

kozzztik commented:
Merged with master again, and #26 now here.

kozzztik commented:
Can I do something to get it in?

warsaw commented:
@kozzztik Hi.  I've been traveling but will soon review this for content.  I see some checks are still failing.  These have to be fixed before the branch can be merged.  Do you know what's causing the failures?

kozzztik commented:
@warsaw seems that problem only in code coverage on Python 3.4.

warsaw commented:
@kozzztik No, I think there are legitimate failures on all three Python versions.  Check the Travis details.  It fails on Python 3.4, 3.5, and 3.6 and not for coverage problems.  I'm happy to fix the coverage issues if you can narrow it down to that.

kozzztik commented:
Fixed code style problems and code coverage. I just covered some cases not related to STARTTLS and increased total project code coverage, so now lmit 96% is passed under py34.  Also fixed one test and removed "State" parameter that seems to be copy-pasted from old SMTP implementation which was classic state machine.

warsaw commented on aiosmtpd/tests/test_handlers.py:
```diff
@@ -38,8 +38,7 @@ def setUp(self):
             client = resources.enter_context(SMTP(*self.address))
-            resources.enter_context(
-                patch('aiosmtpd.handlers._format_peer', _format_peer_address))
+            peer = client.sock.getsockname()
```
I don't mind this change, but there are two places in test_handlers.py where we mock `_format_peer`.  I think I'll just fix this up after landing your branch.

warsaw commented on aiosmtpd/tests/test_starttls.py:
```diff
@@ -0,0 +1,216 @@
+from aiosmtpd.smtp import SMTP as SMTPProtocol
+from email.mime.text import MIMEText
+from smtplib import SMTP
+from .test_smtp import Controller as OldController, Sink
```
Generally I prefer absolute imports, but I think I'll just fix this one up after landing the branch.

warsaw commented:
@kozzztik Thanks for these changes, they look great.  I made two very minor style fixes, and I have two small post-landing changes I'd like to make (plus a NEWS entry), but this branch is ready to land.  I want to do one last test locally (to make sure the Mailman 3 trunk still passes with this branch), but other than that, it's ready to go.  Thank you very much for this great contribution.

kozzztik commented:
@warsaw please can you help me? I`m piece of idiot. If forgot to change commiter name while merging it back from my project, and all my commits marked as "kvolkov" instead if "kozzztik <kozzztik@mail.ru>", so Github thinks that that was not "my" commits.  I`m a liittle proud of this PR, and this mistake makes me really upset. I`m not very familiar with git, and seems that, anyway write rights are needed. I`ll be really happy if you can help me to fix that.

warsaw commented:
@kozzztik Oh darn.  I'm happy to try to help if I can, but I don't know how to do this.  I tried [this recipe](http://stackoverflow.com/questions/750172/change-the-author-and-committer-name-and-e-mail-of-multiple-commits-in-git#1320317) but I got conflicts and the rebasing failed.  I'm also not sure whether a repo surgically altered this way and pushed will affect any one else's clone.  But I'm happy to try to help if you can figure out a foolproof way of doing it.

kozzztik commented:
@warsaw thanks for your try, but seems I just must do something more impressive to improve my profile. Sorry for stealing your time for that.
